handlers.go

```diff
@@ -1,24 +1,169 @@
+	errFailedGetState = errors.New("failed to retrieve saved state")
+	errNoAddr         = errors.New("parameter 'addr' not found")
+	errHashWrongSize  = errors.New("given hash is of invalid size")
+)
+
+func formatIndexPage(appURL *url.URL) string {
+	page := indexPage
+	if appURL != nil {
 		page += fmt.Sprintf("\nIt runs the following code: %s\n"+
 			"Use the following tool to verify the enclave: "+
-			"https://github.com/brave-experiments/verify-enclave", appURL)
+			"https://github.com/brave-experiments/verify-enclave", appURL.String())
 	}
 	return page
 }
 
-// getIndexHandler returns an index handler that informs the visitor that this
-// host runs inside an enclave.
-func getIndexHandler(cfg *Config) http.HandlerFunc {
+// rootHandler returns a handler that informs the visitor that this host runs
+// inside an enclave.  This is useful for testing.
+func rootHandler(cfg *Config) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, formatIndexPage(cfg.AppURL))
 	}
 }
+
+// reqSyncHandler returns a handler that lets the enclave application request
+// state synchronization, which copies the given remote enclave's state into
+// our state.
+//
+// FIXME: https://github.com/brave/nitriding/issues/44
+func reqSyncHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		q := r.URL.Query()
+		// The 'addr' parameter must have the following form:
+		// https://example.com:8443
+		addrs, ok := q["addr"]
+		if !ok {
+			http.Error(w, errNoAddr.Error(), http.StatusBadRequest)
+			return
+		}
+		addr := addrs[0]
+
+		// Are we dealing with a well-formed URL?
+		if _, err := url.Parse(addr); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if err := RequestKeys(addr, e.KeyMaterial); err != nil {
+			http.Error(w, fmt.Sprintf("failed to synchronize state: %v", err), http.StatusInternalServerError)
+			return
+		}
+		w.WriteHeader(http.StatusOK)
+	}
+}
+
+// getStateHandler returns a handler that lets the enclave application retrieve
+// previously-set state.
+func getStateHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/octet-stream")
+		s, err := e.KeyMaterial()
+		if err != nil {
+			http.Error(w, errFailedGetState.Error(), http.StatusInternalServerError)
+			return
+		}
+		n, err := w.Write(s.([]byte))
+		if err != nil {
+			elog.Printf("Error writing state to client: %v", err)
+			return
+		}
+		expected := len(s.([]byte))
+		if n != expected {
+			elog.Printf("Only wrote %d out of %d-byte state to client.", n, expected)
+			return
+		}
+	}
+}
+
+// putStateHandler returns a handler that lets the enclave application set
+// state that's synchronized with another enclave in case of horizontal
+// scaling.  The state can be arbitrary bytes.
+func putStateHandler(e *Enclave) http.HandlerFunc {
```

Comment:
Right, you are probably thinking about the function [respSyncHandler](https://github.com/brave/nitriding/blob/beed0b9606eda4e633330f16bb53d7ff01f4b2f9/keysync_responder.go#L48). The purpose of this endpoint is to let the (trusted) enclave application set whatever state it wants to be set. It's an enclave-internal, trusted endpoint. We allow the application to set arbitrary bytes, so we limit our validation to enforcing a length limit by taking advantage of `newLimitReader`.

Reply:
To isolate public from private endpoints, we are using two separate Web servers: one that binds to localhost (see [here](https://github.com/brave/nitriding/blob/c850d56d687c9df8b6601f150845c4c9c15e9c94/enclave.go#L169)) and another (see [here](https://github.com/brave/nitriding/blob/c850d56d687c9df8b6601f150845c4c9c15e9c94/enclave.go#L165)) that imposes no such restrictions. I'm going to add a comment to the affected endpoint handlers because this is more confusing than it should be.

Reply:
9edb2c4 adds a few comments, explaining that some of our endpoints are enclave-internal.